src/Store/Persistence.re

```diff
@@ -0,0 +1,247 @@
+    key: string,
+    default: 'value,
+    codec: Codec.t('value),
+    get: 'state => 'value,
+    pipe: Pipe.t('value),
+  };
+
+  let define = (key, codec, default, get) => {
+    key,
+    codec,
+    default,
+    get,
+    pipe: Pipe.create(),
+  };
+};
+
+module Store = {
+  type entry('state) =
+    | Entry({
+        definition: Schema.item('state, 'value),
+        mutable value: 'value,
+      })
+      : entry('state);
+
+  let entry = definition => Entry({definition, value: definition.default});
+
+  type t('state) = {
+    name: string,
+    hash: string,
+    entries: list(entry('state)),
+  };
+
+  let instantiate = (name, entries) => {
+    let store = {name, hash: Internal.hash(name), entries: entries()};
+
+    let path =
+      Filesystem.getStoreFolder()
+      |> Result.map(storeFolder => Filename.concat(storeFolder, store.hash))
+      |> ResultEx.flatMap(Filesystem.getOrCreateConfigFolder)
+      |> Result.map(folder => Filename.concat(folder, "store.json"))
+      |> Result.get_ok;
+
+    switch (Yojson.Safe.from_file(path)) {
+    | json =>
+      switch (Json.Decode.(decode_value(key_value_pairs(value), json))) {
+      | Ok(persistedEntries) =>
+        List.iter(
+          (Entry({definition, _} as entry)) =>
+            switch (List.assoc_opt(definition.key, persistedEntries)) {
+            | Some(value) =>
+              switch (
+                Json.Decode.decode_value(definition.codec.decode, value)
+              ) {
+              | Ok(value) => entry.value = value
+              | Error(error) =>
+                let message = Json.Decode.string_of_error(error);
+                Log.error("Error decoding store file: " ++ message);
+                entry.value = definition.default;
+              }
+            | None => entry.value = definition.default
+            },
+          store.entries,
+        )
+
+      | Error(error) =>
+        let message = Json.Decode.string_of_error(error);
+        Log.error("Error parsing store file: " ++ message);
+      }
+    | exception (Sys_error(message)) =>
+      // Most likely because the file doesn't exist, which is expected, but log it just in case.
+      Log.debug("Unable to read store file: " ++ message)
+    };
+
+    store;
+  };
+
+  let persist = store => {
+    Log.debug("Writing store for " ++ store.name);
+    let entries =
+      store.entries
+      |> List.map((Entry({definition, value})) =>
+           (definition.key, value |> definition.codec.encode)
+         );
+
+    let str = Json.Encode.encode_string(Json.Encode.obj, entries);
+
+    let path =
+      Filesystem.getStoreFolder()
+      |> Result.map(storeFolder => Filename.concat(storeFolder, store.hash))
+      |> ResultEx.flatMap(Filesystem.getOrCreateConfigFolder)
+      |> Result.map(folder => Filename.concat(folder, "store.json"))
+      |> Result.get_ok;
+
+    let outChannel = open_out(path);
+    Printf.fprintf(outChannel, "%s", str);
+    close_out(outChannel);
```

Comment:
TODO: This should use Luv to avoid blocking I/O

Reply:
We should look at this across all our File I/O (like the extension scanning, too!)

Reply:
This is actually a bit more involved than I first thought, as it uses `Filesystem` just above here to get the right path. And that logic is so convoluted that I'd really rather not touch it. I guess I'll just duplicate the parts that are blocking here.

Reply:
Or just do it once during instantiation and cache it. The read needs to be blocking anyway.

Reply:
Seems reasonable - I need to profile startup perf and see what the bottlenecks are. Might be some targeted places we could make async (like extension scanning / loading), that would be more impactful.

Reply:
> Or just do it once during instantiation and cache it. The read needs to be blocking anyway.

Yeah, it seems like reading persistence has to be sync on startup, so we know the parameters to pass to the window.